firedrake/mg/interface.py

```diff
@@ -3,22 +3,29 @@
 import firedrake
 from firedrake import ufl_expr
 from firedrake.petsc import PETSc
+from ufl.duals import is_dual
 from . import utils
 from . import kernels
 
 
 __all__ = ["prolong", "restrict", "inject"]
 
 
-def check_arguments(coarse, fine):
+def check_arguments(coarse, fine, needs_dual=False):
+    if is_dual(coarse) != needs_dual:
```

Comment:
@dham This should address point (2/2) from the review. Now we are type checking using `is_dual`.